### PR TITLE
net: tc: Allow user to select pre-emptive threading for RX/TX

### PR DESCRIPTION
**Pull request by jukkar**
Let user to decide whether the RX/TX threads are run in either
co-operative or pre-emptive thread priority.
Default is co-operative threading.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

### CODE REVIEW

**Comment by jukkar** [comment]
@armandciejak  we discussed this in networking forum. This PR is just experimenting and needs more testing before it is ready to upstream.

**Comment by andrewboie** [comment]
What's the functional implication of this? In other words, why would an application want to enable this feature, or leave it disabled?

**Comment by andrewboie** [comment]
If this is related to our conversation about SMP the other day, it's worth noting the co-op threads maintain their individual scheduling guarantees even in an SMP environment. It's just that if you implement a critical section between two threads, relying on one thread having a higher co-op priority than another to guarantee exclusion, that the other thread could be running simultaneously on another CPU, and exclusion guarantees fall over.

**Comment by jukkar** [comment]
> What's the functional implication of this? In other words, why would an application want to enable this feature, or leave it disabled?

This was related to #27662 where incoming network traffic can starve the system. By running RX/TX threads in pre-emptive priority, some other high priority thread is allowed to run.

**Comment by andrewboie** [comment]
> > What's the functional implication of this? In other words, why would an application want to enable this feature, or leave it disabled?
> 
> This was related to #27662 where incoming network traffic can starve the system. By running RX/TX threads in pre-emptive priority, some other high priority thread is allowed to run.

ahh understood thanks!

File: subsys/net/ip/Kconfig [hunk 1]
```diff
@@ -167,6 +167,26 @@ config NET_TC_RX_COUNT
...
+	  Please select the RX/TX threads to be either pre-emptive or
+	  co-operative.
+
+config NET_TC_THREAD_COOPERATIVE
+	bool "Use co-operative TX/RX threads"
+	depends on COOP_ENABLED
+	help
+	  With co-operative threads, the thread cannot be pre-empted.
+
+config NET_TC_THREAD_PREEMPTIVE
```
Comment (rlubos) [inline subsys/net/ip/Kconfig:182]: Should this variant me marked `[EXPERIMENTAL]` perhaps for now?

**Review by rlubos — COMMENTED** [review]
I've tested a few samples with preemptive threads, `lwm2m_client` worked fine, same as `echo_client`/`echo_server` on OpenThread. But when I tried to run `echo_client/server` against each other on qemu_x86, I've ended up with the following:
image
Looked like the client tried to send data before the TCP connection was complete. Not sure what's the root cause of it.

File: subsys/net/ip/Kconfig (hunk 1 above)
Comment (jukkar) [reply subsys/net/ip/Kconfig:182]: > Should this variant me marked `[EXPERIMENTAL]` perhaps for now?

Yes, I will add that.

**Comment by jukkar** [comment]
> But when I tried to run `echo_client/server` against each other on qemu_x86, I've ended up with the following

Thanks for the info, I did not try this combo in my testing but will check it next.

**Comment by jukkar** [comment]
> Looked like the client tried to send data before the TCP connection was complete. Not sure what's the root cause of it.

The reason was that `main()` was running in preemptive priority 0, and network TX/RX thread in priority 8. After setting the main to run in priority 8, the connectivity was ok. This is yet another "confusion" point for the application developers, the thread priorities need to be selected carefully in order to let the certain threads run in correct time. We will probably see lot of issues created for this in the future.

**Comment by jukkar** [comment]
Changes:
* adjust the main() priority to be 8 in echo-client

**Comment by AntoineZen** [comment]
I d'like to jump in here. We are also hit by the issue, and by a case of "Looked like the client tried to send data before the TCP connection was complete". I was quite confuse by the following:

> This is yet another "confusion" point for the application developers, the thread priorities need to be selected carefully in order to let the certain threads run in correct time.

IMO, in a RTOS with preemptive task, this should not be necessary. They are synchronization primitive, and a thread must use them (encapsulated in the socket API) to wait for the network thread to be ready if required. In an RTOS this could involve priory inheritance. Changing thread priority to solve such problem seems a poor to me. Put a semaphore or another barrier to block the application until the connection is completed.

Also, application designer should be able to select the network thread priority depending on they requirement.

**Comment by jukkar** [comment]
> Put a semaphore or another barrier to block the application until the connection is completed.

Sure, that is a good solution. I was just trying to get the existing networking samples to work with minimal changes to them.

**Comment by jukkar** [comment]
> Also, application designer should be able to select the network thread priority depending on they requirement.

Yes, allowing user to select the exact thread priority is still missing here. The situation is a bit more complex as we can have max. 8 RX or TX threads, and one would need to select the base priority correctly. That is the reason why the base priority is set to 8 in the current commits.

**Comment by jukkar** [comment]
* Fixed merge conflict
* Fixed thread priorities in samples and net
* Refactored the RX/TX thread priorities (see the commit `net: tc: Refactor RX and TX thread priorities` for details)

**Comment by jukkar** [comment]
* Fixed the unit tests

**Comment by jukkar** [comment]
Hi all! I am hoping we could get this one merged soonish so I am asking reviews some help to review this. Note that pre-emptive priority support for networking thread it is marked as experimental and is turned off by default. So there should be minimal risk here, and the sooner we merge this, the better the testing coverage by the users there will be.

**Review by rlubos — APPROVED** [review]
Echo samples with preemptive threads work fine now.

**Review by alexanderwachter — APPROVED** [review]
+1 for 6locan

**Pull request merged**